### ISSUE TITLE
libgcrypt 1.8.6 built on macOS 10.15

### ISSUE THREAD

**Issue by droidmonkey** [Homebrew/homebrew-core#63267]
# Bug report

**Please note we will close your issue without comment if you delete, do not read or do not fill out the issue checklist below and provide ALL the requested information. If you repeatedly fail to use the issue template, we will block you from ever submitting issues to Homebrew again.**

- [x] ran `brew update` and can still reproduce the problem?
- [x] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [ ] ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?
- [ ] if `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?

## What you were trying to do (and why)

Use libgcrypt 1.8.6 with KeePassXC to deploy to macOS versions 10.13+

## What happened (include command output)

```
Dyld Error Message:
Symbol not found: ____chkstk_darwin
Referenced from: /Applications/KeePassXC.app/Contents/MacOS/../Frameworks/libgcrypt.20.dylib (which was built for Mac OS X 10.15)
Expected in: /usr/lib/libSystem.B.dylib
```

## What you expected to happen

The libgcrypt dylib should have been built to support macOS 10.13+. The previous version of libgcrypt (1.8.5) works properly.

## Step-by-step reproduction instructions (by running `brew install` commands)

1. Build KeePassXC
2. Install on macOS 10.14 or below
3. See the error

**Comment by SMillerDev** [comment]
It's a library inside keepassxc that references a system file. I'm not sure how this is related to homebrew.

**Comment by phoerious** [comment]
I have to debug it more closely on a Mac (which I don't have at hand right now), but from the error message it looks like libgcrypt.20.dylib tries to use a symbol, which doesn't exist on Mojave. libgcrypt.20.dylib in KeePassXC comes from Homebrew.

Now the install formula looks like it is compiled on-the-fly at the start of the install process, so if the package is installed on Catalina, it is going to be compatible only with Catalina. The install formula should set an explicit target release to avoid that.

**Comment by SMillerDev** [comment]
Why is it copying shared libraries from Homebrew? That's a very strange tactic.

For formulae we set the target that we're compiling on for our bottles, running it on another OS is not supported.

**Comment by phoerious** [comment]
How else would you ship a program if not with all needed libraries? macdeployqt automatically collects all needed libraries and bundles them inside the application folder.

**Comment by SMillerDev** [comment]
You compile them as part of the program to make sure that it matches all your requirement, Homebrew formulae really aren't made to be just copied around, they are made to work within Homebrew, by Homebrew.

**Comment by phoerious** [comment]
I am not copying the formula, I am only copying the built libraries and making those compatible is as easy as setting `MACOSX_DEPLOYMENT_TARGET` in 99% of the cases. I fail to see how this is an uncommon workflow. We do it for KeePassXC itself as well, otherwise we'd need one separate build system and binary for every macOS release. The only alternative I see is compiling all libraries by hand and replicating exactly what brew is doing except for setting the target to 10.14, which is the earliest release that Qt 5.15 supports and not unreasonably old.

**Comment by SMillerDev** [comment]
This is not what Homebrew is meant for. It's meant to install software on a single system and we provide no support for any copying or compatibility of files.

> The only alternative I see is compiling all libraries by hand and replicating exactly what brew is doing except for setting the target to 10.13, which is the earliest release that Qt 5.15 supports and not unreasonably old.

That is unfortunately the way it is in all software. If you need shared libraries, you need to make sure that the user has them installed (which macOS doesn't have a mechanism for). Otherwise you need to make sure that the software bundles all required parts, this is what software like snapcraft became popular for. 

---

TL;DR this is not supported, closing

**Comment by phoerious** [comment]
I suppose we have to find another solution then and dump Homebrew altogether. We cannot deal with Apple's short support cycles AND Homebrew's even shorter support cycles, which do not even include Qt LTS releases.

> If you need shared libraries, you need to make sure that the user has them installed (which macOS doesn't have a mechanism for).

Except for Homebrew. Ironic, isn't it?

**Comment by SMillerDev** [comment]
> Except for Homebrew. Ironic, isn't it?

But since you can't expect Homebrew to be installed as you can expect `apt` or `yum`, it doesn't really. And even then you can't expect something distributed outside of Homebrew to demand Homebrew dependencies.

> I would only expect such a trivial no-cost best best-effort fix to at least be considered.

This isn't a fix though, this is an issue opened mentioning an unsupported use case.

**Comment by phoerious** [comment]
> And even then you can't expect something distributed outside of Homebrew to demand Homebrew dependencies.

KeePassXC comes with a consistent set of bundled libraries, which are all built with Homebrew and are guaranteed to work together. We do not expect any Homebrew dependencies that are not bundled with the application, so this statement doesn't apply. The only thing that changes between my build system and the user's system is the set of standard macOS system libraries and thus the only breaking issue here is that some (!) libraries are not built with backwards compatibility in mind. Those that are built upstream and pulled as binaries usually are built that way by their maintainers (such as Qt), those that are compiled by Homebrew during the installation are not. These are the ones that will break as soon as system APIs are used that are not available on older macOS versions.

It is easy to say our use case is not supported, but the fix is trivial and comes at no cost, while at the same time increasing the value of Homebrew massively. I cannot wrap my head around why this would be such a big deal. Every sane person builds distributable macOS binaries with at least the last three macOS releases as target, which are still officially supported by Apple.

**Comment by jonchang** [comment]
This is one of the reasons why we build our binary bottles separately on each supported macOS version, rather than distributing a single binary that works on all three versions.

> I would only expect such a trivial no-cost best best-effort fix to at least be considered.

We do not have the maintainer or continuous integration capacity to do as you're suggesting (to ensure that binaries built on 10.15 work on e.g. 10.13). This is neither a trivial nor a no-cost fix.
